Source/Core/Common/Debug/CodeTrace.cpp

```diff
@@ -0,0 +1,394 @@
+// Copyright 2022 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "Common/Debug/CodeTrace.h"
+
+#include <chrono>
+#include <regex>
+
+#include "Common/Event.h"
+#include "Common/StringUtil.h"
+#include "Core/Debugger/PPCDebugInterface.h"
+#include "Core/HW/CPU.h"
+#include "Core/PowerPC/PowerPC.h"
+
+static bool CompareInstruction(std::string_view instruction, std::vector<std::string> type_compare)
+{
+  for (auto& s : type_compare)
+  {
+    if (StringBeginsWith(instruction, s))
+      return true;
+  }
+
+  return false;
+}
+
+static bool IsInstructionLoadStore(std::string_view ins)
+{
+  return (StringBeginsWith(ins, "l") && !StringBeginsWith(ins, "li")) ||
+         StringBeginsWith(ins, "st") || StringBeginsWith(ins, "psq_l") ||
+         StringBeginsWith(ins, "psq_s");
+}
+
+const u32 GetMemoryTargetSize(const std::string& instr)
+{
+  // Word-size operations are taken as the default, check the others.
+  u32 size = 4;
+
+  const std::string op = instr.substr(0, 4);
+  const char byte = 'b';
+  const char half = 'h';
+  const char double_word = 'd';
+  const char paired = 'p';
+
+  // Actual range is 0 to size - 1;
+  if (op.find(byte) != std::string::npos)
+    size = 1;
+  else if (op.find(half) != std::string::npos)
+    size = 2;
+  else if (op.find(double_word) != std::string::npos || op.find(paired) != std::string::npos)
+    size = 8;
+
+  return size;
+}
+
+const std::tuple<std::vector<std::string>, std::set<u32>> CodeTrace::GetTracked()
+{
+  // Output without conversions, incase the u32 memory address is to be used.
+  return std::make_tuple(m_reg_autotrack, m_mem_autotrack);
+}
+
+void CodeTrace::SetRegTracked(const std::string reg)
+{
+  m_reg_autotrack.push_back(reg);
+}
+
+const InstructionAttributes CodeTrace::GetInstructionAttributes(TraceOutput instruction) const
+{
+  // Slower process of breaking down saved instruction. Only used when stepping through code if a
+  // decision has to be made, otherwise used afterwards on a log file.
+  InstructionAttributes tmp_attributes;
+  tmp_attributes.instruction = instruction.instruction;
+  tmp_attributes.address = PC;
+  std::string instr = instruction.instruction;
+
+  // Convert sp, rtoc, and ps to r1, r2, and F#. ps is handled like a float operation.
+  std::regex replace_sp("(\\W)sp");
```

Comment:
All these regex expressions should be defined as member variables in the class constructor (possibly even `static` if the `CodeTrace` class is constructed a lot).  Regex is really expensive to create.  You can also create it with the `std::regex::optimize` flag for an ever so tiny boost in perf.

Reply:
Thanks so much for these optimizations!  CodeTrace isn't constructed too often, so I'm not sure they need to be static.